Added Warping and Text Transformation as well as text erasing

longkidkoolstar opened the pull request:
- Implement warp mode and text mask strokes in TextBlock class
- Add warp editing controls in TextBlkShapeControl
- Introduce text mask brush functionality in Canvas
- Add undo/redo support for warp and text mask operations
- Implement quad and mesh warping algorithms in TextBlkItem

Also made pressing escape for Text Erase to escape out of it

longkidkoolstar committed:
feat(textblock): add warp transformation and text mask editing support

- Implement warp mode and text mask strokes in TextBlock class
- Add warp editing controls in TextBlkShapeControl
- Introduce text mask brush functionality in Canvas
- Add undo/redo support for warp and text mask operations
- Implement quad and mesh warping algorithms in TextBlkItem

longkidkoolstar committed:
feat(text-tools): add option to move text tools to sidebar

- Add config option to move warp/transform/eraser tools to advanced text sidebar
- Implement signal connections between canvas and text panel components
- Update context menu to conditionally show text tools based on config
- Add new methods for text tool operations in canvas and text panel

longkidkoolstar committed:
refactor(text formatting): move angle and squeeze actions to context menu

Remove angle and squeeze buttons from text format panel and move them to the context menu for cleaner UI. This change simplifies the text formatting panel while maintaining all functionality through the context menu.

longkidkoolstar committed:
feat(text-warping): enhance text warping with mesh mode and center control

- Add mesh-based warping mode for more flexible text deformation
- Implement center control point for rise/fall warping effects
- Support edge handles for both quad and mesh warping modes
- Improve warping visualization with stroke outlines

longkidkoolstar committed:
perf(textitem): improve text rendering quality with scaling

Add raster scaling to text rendering to improve quality on high DPI displays. The changes include:
- Calculating appropriate scale based on device DPI and dimensions
- Applying scaling to the rendered image
- Adding antialiasing and smooth transform hints
- Ensuring proper image bounds when drawing

longkidkoolstar committed:
feat(render): add high quality warp preview option

Add configurable option to enable high quality rendering for text warp preview. When enabled, uses higher raster scale for better quality. Also adds cache invalidation method to update text items when quality setting changes.

dmMaze commented:
Well done and thank you for your hard work! 

The warping seems to result in the following slashed artifacts: 
<img width="520" height="311" alt="image" src="https://github.com/user-attachments/assets/49ada054-7d34-49da-a7f0-c762a5b45a83" />
The font is Microsoft YaHei UI, many other fonts (such as Arial) also share the issue.

Besides, the corner ctrl blocks' movement seems to be attached to edge blocks: 
![output](https://github.com/user-attachments/assets/ae362dfa-f4a6-4570-9cac-40096e4e94e4)

longkidkoolstar committed:
feat(canvas): dynamically calculate mesh size based on item dimensions

Implement adaptive mesh sizing for warp editing based on bounding rect dimensions and target cell size. Add mesh edge point calculation and improved z-value handling for control blocks. Introduce OpenCV-based mesh warping with caching for better performance.

Refactor text alignment handling for vertical text and improve mesh warping quality with better interpolation and edge handling.

longkidkoolstar commented:
> Well done and thank you for your hard work!
> 
> The warping seems to result in the following slashed artifacts: <img alt="image" width="520" height="311" src="https://private-user-images.githubusercontent.com/51270320/528943661-49ada054-7d34-49da-a7f0-c762a5b45a83.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY0MTczMjEsIm5iZiI6MTc2NjQxNzAyMSwicGF0aCI6Ii81MTI3MDMyMC81Mjg5NDM2NjEtNDlhZGEwNTQtN2QzNC00OWRhLWE3ZjAtYzc2MmE1YjQ1YTgzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjIyVDE1MjM0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwMTBiY2I1ZTYzZjYzZTYwNmQyMzczNDI5OTVmMWFmN2FlODgyNTcwNzY3ZTBiNWU1Zjk2MmM3MTRkZDU5NWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jLN-CQGnUGgUNtRe99cR7EdAPYGfZcWcULe5bwUrtDE"> The font is Microsoft YaHei UI, many other fonts (such as Arial) also share the issue.
> 
> Besides, the corner ctrl blocks' movement seems to be attached to edge blocks: ![output](https://private-user-images.githubusercontent.com/51270320/528944656-ae362dfa-f4a6-4570-9cac-40096e4e94e4.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY0MTczMjEsIm5iZiI6MTc2NjQxNzAyMSwicGF0aCI6Ii81MTI3MDMyMC81Mjg5NDQ2NTYtYWUzNjJkZmEtZjRhNi00NTcwLTljYWMtNDAwOTZlNGU5NGU0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjIyVDE1MjM0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3MTQzYTFjZWIzM2Y2M2JjNDUyZWQ4MmUzNzVjZGY3ODA1NDZmNjhkOTcyYjAxODI4ZDIxMzgxMmQ3OTZmYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Cg_UIPMfz7_XUrSNDkPNjxFUei7FqT8zHwIBEL2KwCk) [ ![output](https://private-user-images.githubusercontent.com/51270320/528944656-ae362dfa-f4a6-4570-9cac-40096e4e94e4.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY0MTczMjEsIm5iZiI6MTc2NjQxNzAyMSwicGF0aCI6Ii81MTI3MDMyMC81Mjg5NDQ2NTYtYWUzNjJkZmEtZjRhNi00NTcwLTljYWMtNDAwOTZlNGU5NGU0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjIyVDE1MjM0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3MTQzYTFjZWIzM2Y2M2JjNDUyZWQ4MmUzNzVjZGY3ODA1NDZmNjhkOTcyYjAxODI4ZDIxMzgxMmQ3OTZmYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Cg_UIPMfz7_XUrSNDkPNjxFUei7FqT8zHwIBEL2KwCk) ](https://private-user-images.githubusercontent.com/51270320/528944656-ae362dfa-f4a6-4570-9cac-40096e4e94e4.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY0MTczMjEsIm5iZiI6MTc2NjQxNzAyMSwicGF0aCI6Ii81MTI3MDMyMC81Mjg5NDQ2NTYtYWUzNjJkZmEtZjRhNi00NTcwLTljYWMtNDAwOTZlNGU5NGU0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjIyVDE1MjM0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3MTQzYTFjZWIzM2Y2M2JjNDUyZWQ4MmUzNzVjZGY3ODA1NDZmNjhkOTcyYjAxODI4ZDIxMzgxMmQ3OTZmYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Cg_UIPMfz7_XUrSNDkPNjxFUei7FqT8zHwIBEL2KwCk) [ ](https://private-user-images.githubusercontent.com/51270320/528944656-ae362dfa-f4a6-4570-9cac-40096e4e94e4.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY0MTczMjEsIm5iZiI6MTc2NjQxNzAyMSwicGF0aCI6Ii81MTI3MDMyMC81Mjg5NDQ2NTYtYWUzNjJkZmEtZjRhNi00NTcwLTljYWMtNDAwOTZlNGU5NGU0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjIyVDE1MjM0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3MTQzYTFjZWIzM2Y2M2JjNDUyZWQ4MmUzNzVjZGY3ODA1NDZmNjhkOTcyYjAxODI4ZDIxMzgxMmQ3OTZmYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Cg_UIPMfz7_XUrSNDkPNjxFUei7FqT8zHwIBEL2KwCk)

Thanks for the feedback!

Artifacts: Fixed. I rewrote the warp backend to use continuous remapping, which completely eliminates the diagonal seams/slashes.

Handles: Fixed. I updated the control logic to decouple the corners from the edges. Moving an edge now leaves the corner handle in place, but I left moving the corner affect the edge for a more unified wraping. If you think itll be better I can remove that one too.

![warping](https://github.com/user-attachments/assets/9ee9dc77-347b-4316-a7e6-f0bef69566f1)
![erasing](https://github.com/user-attachments/assets/0d6c090e-6d51-4bb8-9a5f-c9559385e1b3)

dmMaze commented:
Here's another bug involving CJK characters: 
![output](https://github.com/user-attachments/assets/61376a29-0811-4fbb-9514-4916da615fe2)

I typed "测试" and warping it resulted in an abrupt shift for the central handles

dmMaze commented:
And the handle won't disappear while creating a rect region (which it should) for the rectangle tool in painting mode.

longkidkoolstar committed:
refactor(texteditshapecontrol): improve control visibility handling

Add _controls_hidden flag to track control visibility state and consolidate hide/show logic in hideControls/showControls methods. Remove duplicate visibility control code from startEditing/endEditing methods.

longkidkoolstar commented:
> And the handle won't disappear while creating a rect region (which it should) for the rectangle tool in painting mode.

I fixed the issues with the rect region for the rectangle tool in painting mode, but for the other 
issue with the abrupt shift in the central handles I was unable to recreate it ![middle](https://github.com/user-attachments/assets/ed894f51-08b6-4327-87a9-eb98eb4b6d24)

Could you perhaps mention the font you were using?

dmMaze commented:
@longkidkoolstar 
All fonts I'm using exhibit the issue. You could try SimSun or Noto Sans CJK SC, also try to make the font size larger, and drag & drop the central handle several times

dmMaze commented:
Besides, please make an option to disable the warping effect, remind me if it already exists.
And is there a good reason that the central handle can't move horizontally?

longkidkoolstar committed:
feat(text-warp): add rise/fall control for text warp mesh

- Add warp_rise_fall_u and warp_rise_fall_amp fields to TextBlock
- Implement interactive dragging for rise/fall control point
- Improve mesh sampling with bilinear interpolation
- Add preset handling for arc_up and arc_down warp modes
